src/string_bytes.cc

```diff
@@ -72,10 +73,15 @@ class ExternString: public ResourceType {
       ExternString* h_str = new ExternString<ResourceType, TypeName>(isolate,
                                                                      data,
                                                                      length);
-      Local<String> str = String::NewExternal(isolate, h_str);
+      MaybeLocal<String> str = String::NewExternal(isolate, h_str);
       isolate->AdjustAmountOfExternalAllocatedMemory(h_str->byte_length());
 
-      return scope.Escape(str);
+      if (str.IsEmpty()) {
+        delete h_str;
+        return Local<String>();
+      }
```

Comment:
Using `new TypeName[length]` is a bug. If the machine is OOM then you're screwed. Here's an example script:

```js
var a = [];
var f = false;

while (!f) {
  try {
    a.push(new Buffer(1024 * 1024 * 1024).fill('a'));
  } catch (e) {
    f = true;
  }
}

console.log(a[a.length - 1].length);      // 1073741824
console.log(a[a.length - 1].toString());  // undefined
```

So `NewFromCopy()` should use `malloc()`, check if `ptr == nullptr` and return `Local<Object>()` if that's the case. Of course proper error handling will need to be in place.

This will account for whether `kStringMaxLength` was exceeded and if the memory ran out of memory while creating the string.

Note: This is completely different from how v2.x handles things now.